internal/venice-common/src/main/java/com/linkedin/venice/grpc/GrpcUtils.java

```diff
@@ -92,4 +94,19 @@ private static KeyStore loadStore(String path, char[] password, String type)
     }
     return keyStore;
   }
+
+  /**
+   * Converts a Netty ByteBuf to a ByteString, checking if it has a backing array to avoid manual copying.
+   *
+   * @param body The ByteBuf to be converted to ByteString.
+   * @return The resulting ByteString.
+   */
+  public static ByteString toByteString(ByteBuf body) {
+    if (body.hasArray()) {
+      // Directly use the backing array to avoid copying
+      return ByteString.copyFrom(body.array(), body.arrayOffset() + body.readerIndex(), body.readableBytes());
+    }
+    // Fallback to nioBuffer() to handle the conversion efficiently
+    return ByteString.copyFrom(body.nioBuffer());
```

Comment:
I see your point. 

>The more efficient approach might be to do response streaming, but I'm not sure what that entails.

That would be ideal and I do have plan to add true streaming API. But for now if you still think the cost of copying CompositeByteBuf is substantial then maybe we'll have to expand the scope to accomodate this change. 


> since you are refactoring the protocol buffer schemas anyway, might be to model the response as a collection rather than one big bytes field

Yeah that's also possible.

Reply:
Dropping this comment for now, since the solution would require too much changes and this PR is already big.